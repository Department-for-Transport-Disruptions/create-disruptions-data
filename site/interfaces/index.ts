import { NextApiRequest } from "next";
import { z } from "zod";
import { ServerResponse } from "http";
import { Service, Stop } from "../schemas/consequence.schema";
import { ConsequenceType } from "../schemas/type-of-consequence.schema";

export interface ErrorInfo {
    errorMessage: string;
    id: string;
    userInput?: string;
}

export interface ResponseWithLocals extends ServerResponse {
    locals: {
        nonce: string;
        csrfToken: string;
    };
}

export interface CookiePolicy {
    essential: boolean;
    usage: boolean;
}

export interface CookiesApiRequest extends NextApiRequest {
    body: {
        tracking: "on" | "off";
    };
}

export interface FormBase<T> {
    value?: string;
    display: string;
    displaySize?: "s" | "m" | "l" | "xl";
    inputName: Extract<keyof T, string>;
    initialErrors?: ErrorInfo[];
    stateUpdater: (change: string, field: keyof T) => void;
    schema?: z.ZodTypeAny;
}

export interface DisplayValuePair<T = string> {
    display: string;
    value: T;
    checked?: boolean;
}

export interface Feedback {
    question: string;
    answer: string;
}

export interface TestInputs {
    field1: string;
    field2: boolean;
    field3: string;
    field4: string;
}

export interface PageState<T> {
    errors: ErrorInfo[];
    inputs: T;
}

export interface SocialMediaPost {
    messageToAppear: string;
    publishDate: string;
    publishTime: string;
    accountToPublish: string;
}

export interface CreateConsequenceProps<T> {
    initialPageState: PageState<Partial<T>>;
    previousConsequenceInformation: ConsequenceType;
    initialServices?: Service[];
    initialStops?: Stop[];
    allStops?: Stop[]
}
