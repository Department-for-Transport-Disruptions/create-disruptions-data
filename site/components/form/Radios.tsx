import kebabCase from "lodash/kebabCase";
import { Fragment, ReactElement, useEffect, useRef, useState } from "react";
import FormElementWrapper, { FormGroupWrapper } from "./FormElementWrapper";
import { DisplayValuePair, ErrorInfo, FormBase } from "../../interfaces";
import { handleBlur } from "../../utils/formUtils";

interface RadiosProps<T> extends FormBase<T> {
    radioDetail: DisplayValuePair[];
    paddingTop?: number;
}

const Radios = <T extends object>({
    display,
    displaySize = "s",
    inputName,
    radioDetail,
    value,
    initialErrors = [],
    stateUpdater,
    paddingTop,
    schema,
}: RadiosProps<T>): ReactElement => {
    const [errors, setErrors] = useState<ErrorInfo[]>(initialErrors);
    const inputId = kebabCase(inputName);

    return (
        <FormGroupWrapper errorIds={[inputName]} errors={errors}>
            <fieldset className="govuk-fieldset" id={inputId}>
                <legend className={`govuk-fieldset__legend${paddingTop ? ` govuk-!-padding-top-${paddingTop}` : ""}`}>
                    <span className={`govuk-heading-${displaySize} govuk-!-margin-bottom-0`}>{display}</span>
                </legend>
                <FormElementWrapper errors={errors} errorId={inputName} errorClass="govuk-radios--error">
                    <div className="govuk-radios" data-module="govuk-radios">
                        {radioDetail.map((input, index) => (
                            <Fragment key={`radio-${input.value}`}>
                                <div
                                    className={`govuk-radios__item${
                                        index < radioDetail.length - 1 ? " govuk-!-margin-bottom-1" : ""
                                    }`}
                                >
                                    <input
                                        className="govuk-radios__input"
                                        id={`${inputId}-${input.value}`}
                                        name={inputName}
                                        type="radio"
                                        value={input.value}
                                        onBlur={(e) =>
                                            handleBlur(e.target.value, inputName, stateUpdater, setErrors, schema)
                                        }
                                        onChange={(e) => stateUpdater(e.currentTarget.value, inputName)}
                                        defaultChecked={input.value === value}
                                        data-aria-controls={
                                            input.conditionalElement ? `${inputId}-${input.value}-conditional` : null
                                        }
                                    />
                                    <label
                                        className="govuk-label govuk-radios__label"
                                        htmlFor={`${inputId}-${input.value}`}
                                    >
                                        {input.display}
                                    </label>
                                </div>
                                {input.conditionalElement ? (
                                    <div
                                        className="govuk-radios__conditional govuk-radios__conditional--hidden"
                                        id={`${inputId}-${input.value}-conditional`}
                                    >
                                        <div className="govuk-form-group">{input.conditionalElement}</div>
                                    </div>
                                ) : null}
                            </Fragment>
                        ))}
                    </div>
                </FormElementWrapper>
            </fieldset>
        </FormGroupWrapper>
    );
};

export default Radios;
