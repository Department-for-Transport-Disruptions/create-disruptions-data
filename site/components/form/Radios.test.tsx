import renderer from "react-test-renderer";
import { describe, it, expect, vi } from "vitest";
import Radios from "./Radios";
import { TestInputs } from "../../interfaces";

describe("Radios", () => {
    it("should render correctly with no errors", () => {
        const tree = renderer
            .create(
                <Radios<TestInputs>
                    display="Type of disruption"
                    inputId="field1"
                    radioDetail={[
                        {
                            value: "planned",
                            display: "Planned",
                        },
                        {
                            value: "unplanned",
                            display: "Unplanned",
                        },
                    ]}
                    inputName="typeOfDisruption"
                    stateUpdater={vi.fn()}
                    value={""}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with errors", () => {
        const tree = renderer
            .create(
                <Radios<TestInputs>
                    display="Type of disruption"
                    inputId="field1"
                    initialErrors={[{ errorMessage: "There was an error", id: "field1" }]}
                    radioDetail={[
                        {
                            value: "planned",
                            display: "Planned",
                        },
                        {
                            value: "unplanned",
                            display: "Unplanned",
                        },
                    ]}
                    inputName="typeOfDisruption"
                    stateUpdater={vi.fn()}
                    value={""}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with a padding", () => {
        const tree = renderer
            .create(
                <Radios<TestInputs>
                    display="Type of disruption"
                    inputId="field1"
                    radioDetail={[
                        {
                            value: "planned",
                            display: "Planned",
                        },
                        {
                            value: "unplanned",
                            display: "Unplanned",
                        },
                    ]}
                    inputName="typeOfDisruption"
                    stateUpdater={vi.fn()}
                    value={""}
                    paddingTop={2}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
