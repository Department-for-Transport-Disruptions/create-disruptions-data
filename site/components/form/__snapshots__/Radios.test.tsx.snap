// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Radios > should render correctly with a padding 1`] = `
<div
  className="govuk-form-group"
>
  <fieldset
    className="govuk-fieldset"
    id="field-1"
  >
    <legend
      className="govuk-fieldset__legend govuk-!-padding-top-2"
    >
      <span
        className="govuk-heading-s govuk-!-margin-bottom-0"
      >
        Type of disruption
      </span>
    </legend>
    <div
      className=""
    >
      <div
        className="govuk-radios"
      >
        <div
          className="govuk-radios__item govuk-!-margin-bottom-1"
        >
          <input
            className="govuk-radios__input"
            defaultChecked={false}
            id="field-1-planned"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            type="radio"
            value="planned"
          />
          <label
            className="govuk-label govuk-radios__label"
            htmlFor="field-1-planned"
          >
            Planned
          </label>
        </div>
        <div
          className="govuk-radios__item"
        >
          <input
            className="govuk-radios__input"
            defaultChecked={false}
            id="field-1-unplanned"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            type="radio"
            value="unplanned"
          />
          <label
            className="govuk-label govuk-radios__label"
            htmlFor="field-1-unplanned"
          >
            Unplanned
          </label>
        </div>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`Radios > should render correctly with errors 1`] = `
<div
  className="govuk-form-group govuk-form-group--error"
>
  <fieldset
    className="govuk-fieldset"
    id="field-1"
  >
    <legend
      className="govuk-fieldset__legend"
    >
      <span
        className="govuk-heading-s govuk-!-margin-bottom-0"
      >
        Type of disruption
      </span>
    </legend>
    <div
      className=""
    >
      <span
        className="govuk-error-message"
        id="field1-error"
      >
        <span
          className="govuk-visually-hidden"
        >
          Error: 
        </span>
        There was an error
      </span>
      <div
        aria-describedby="field1-error"
        className="govuk-radios govuk-radios--error"
      >
        <div
          className="govuk-radios__item govuk-!-margin-bottom-1"
        >
          <input
            className="govuk-radios__input"
            defaultChecked={false}
            id="field-1-planned"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            type="radio"
            value="planned"
          />
          <label
            className="govuk-label govuk-radios__label"
            htmlFor="field-1-planned"
          >
            Planned
          </label>
        </div>
        <div
          className="govuk-radios__item"
        >
          <input
            className="govuk-radios__input"
            defaultChecked={false}
            id="field-1-unplanned"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            type="radio"
            value="unplanned"
          />
          <label
            className="govuk-label govuk-radios__label"
            htmlFor="field-1-unplanned"
          >
            Unplanned
          </label>
        </div>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`Radios > should render correctly with no errors 1`] = `
<div
  className="govuk-form-group"
>
  <fieldset
    className="govuk-fieldset"
    id="field-1"
  >
    <legend
      className="govuk-fieldset__legend"
    >
      <span
        className="govuk-heading-s govuk-!-margin-bottom-0"
      >
        Type of disruption
      </span>
    </legend>
    <div
      className=""
    >
      <div
        className="govuk-radios"
      >
        <div
          className="govuk-radios__item govuk-!-margin-bottom-1"
        >
          <input
            className="govuk-radios__input"
            defaultChecked={false}
            id="field-1-planned"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            type="radio"
            value="planned"
          />
          <label
            className="govuk-label govuk-radios__label"
            htmlFor="field-1-planned"
          >
            Planned
          </label>
        </div>
        <div
          className="govuk-radios__item"
        >
          <input
            className="govuk-radios__input"
            defaultChecked={false}
            id="field-1-unplanned"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            type="radio"
            value="unplanned"
          />
          <label
            className="govuk-label govuk-radios__label"
            htmlFor="field-1-unplanned"
          >
            Unplanned
          </label>
        </div>
      </div>
    </div>
  </fieldset>
</div>
`;
